modules/gapi/src/compiler/passes/transformations.cpp

```diff
@@ -0,0 +1,159 @@
+    const auto size = transforms.size();
+    if (0u == size) return;
+    GAPI_Assert(0u == patterns.size() || size == patterns.size());
+    patterns.resize(size);
+    std::vector<std::unique_ptr<ade::Graph>> substitutes(size);
+
+    // 1. pre-generate all required graphs
+    for (auto it : ade::util::zip(ade::util::toRange(transforms),
+                                  ade::util::toRange(patterns),
+                                  ade::util::toRange(substitutes))) {
+        const auto& t = std::get<0>(it);
+        auto& p = std::get<1>(it);
+        auto& s = std::get<2>(it);
+
+        p = makeGraph(t.pattern());  // NB: these patterns are re-used in apply_transformation pass
+        s = makeGraph(t.substitute());
+    }
+
+    // FIXME: verify other types of endless loops - are there any other, though?
+
+    const auto empty = [] (const SubgraphMatch& m) {
+        return m.inputDataNodes.empty() && m.startOpNodes.empty()
+            && m.finishOpNodes.empty() && m.outputDataNodes.empty()
+            && m.inputTestDataNodes.empty() && m.outputTestDataNodes.empty();
+    };
+    // 2. verify there are no patterns in substitutes
+    for (size_t i = 0; i < size; ++i) {
+        auto& p = patterns[i];
+        for (size_t j = i; j < size; ++j) {
+            auto& s = substitutes[j];
+
+            auto matchInSubstitute = findMatches(*p, *s);
+            if (!empty(matchInSubstitute)) {
+                std::stringstream ss;
+                ss << "Error: pattern (from transformation #" << i << ") detected inside substitute"
+                      " (from transformation #" << j << ")";
+                throw std::runtime_error(ss.str());
+            }
+        }
+    }
+}
+
+void applyTransformations(ade::passes::PassContext& ctx,
+                          const gapi::GKernelPackage& pkg,
+                          const std::vector<std::unique_ptr<ade::Graph>>& patterns)
+{
+    const auto& transforms = pkg.get_transformations();
+    const auto size = transforms.size();
+    if (0 == size) return;
+    // Note: patterns are already generated at this point
+    GAPI_Assert(patterns.size() == transforms.size());
+
+    // transform as long as it is possible
+    // Note: check_transformations pass must handle endless loops and such
+    bool canTransform = true;
+    while (canTransform)
+    {
+        canTransform = false;
+
+        // iterate through every transformation and try to transform graph parts
+        for (auto it : ade::util::zip(ade::util::toRange(transforms), ade::util::toRange(patterns)))
+        {
+            const auto& t = std::get<0>(it);
+            auto& p = std::get<1>(it);  // Note: using pre-created graphs
+            GAPI_Assert(nullptr != p);
+
+            // if at least one transformation happend, it is possible that other transformations
+            // will be applied in the next "round"
+            canTransform |= tryToSubstitute(ctx.graph, p, t.substitute());
```

Comment:
if substitution was successful here, this `for` loop continues while it should restart